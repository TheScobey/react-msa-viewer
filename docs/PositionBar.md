# `PositionBar` (component)

Creates a PositionBar of markers for every n-th sequence column.

## Props

### `engine`

Rendering engine: `canvas` or `webgl` (experimental).

type: `enum('canvas'|'webgl')`
defaultValue: `"canvas"`


### `font`

Font of the sequence labels, e.g. `20px Arial`

type: `string`
defaultValue: `"12px Arial"`


### `height`

Width of the component (in pixels), e.g. `100`

type: `number`
defaultValue: `100`


### `markerSteps`

At which steps the position labels should appear, e.g. `2` for (1, 3, 5)

type: `number`
defaultValue: `2`


### `msecsPerFps`

defaultValue: `60`


### `msecsPerSecs`

Maximum number of frames per second, e.g. `1000 / 60`

type: `number`


### `startIndex`

At which number the PositionBar marker should start counting.
Typical values are: `1` (1-based indexing) and `0` (0-based indexing).

type: `number`
defaultValue: `1`


### `style`

Custom style configuration.

type: `object`


### `width`

Width of the component (in pixels), e.g. `100`

type: `number`
defaultValue: `100`

