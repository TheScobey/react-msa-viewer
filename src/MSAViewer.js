import React, { Component } from 'react';
import PropTypes from 'prop-types';

const schemes = new (require('msa-colorschemes'))();

class MSAViewer extends Component {

  constructor(props) {
    super(props);
    this.canvas = React.createRef();
  }

  componentDidMount() {
    this.ctx = this.canvas.current.getContext('2d');
    this.scheme = schemes.getScheme(this.props.scheme);
    this.drawSequences();
  }

  drawSequences() {
    const xSize = 20;
    const ySize = 20;
    let yPos = 0, xPos = 0;
    for (const seq of this.props.sequences) {
      for (const el of seq.sequence) {
        this.ctx.font = "20px Arial";
        this.ctx.fillStyle = this.scheme.getColor(el);
        this.ctx.fillRect(xPos, yPos, xSize, ySize);
        this.ctx.fillStyle = "#000000";
        this.ctx.fillText(el, xPos + 2, yPos + 17);
        xPos += xSize;
      }
      xPos = 0;
      yPos += ySize;
    }
  }

  componentDidUpdate() {

  }

  shouldComponentUpdate(newProps) {
    return this.props.target !== newProps.target;
  }

  render() {
    return (
      <canvas ref={this.canvas} with={500} height={500}>
      </canvas>
    );
  }
}

const SequencePropType = PropTypes.shape({
  name: PropTypes.string,
  sequence: PropTypes.string,
})

MSAViewer.propTypes = {
  color: PropTypes.string,
  scheme: PropTypes.string,
  sequences: PropTypes.arrayOf(SequencePropType).isRequired,
};

MSAViewer.defaultProps = {
  color: "red",
  scheme: "clustal",
}

export {MSAViewer};
export default MSAViewer;
