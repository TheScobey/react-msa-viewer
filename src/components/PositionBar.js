/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import msaConnect from '../store/connect'
import PropTypes from 'prop-types';

import { throttle } from 'lodash-es';

import Canvas from '../drawing/canvas';

import createRef from 'create-react-ref/lib/createRef';

/**
TODO:
- buffer to animation frame
- make styling flexible
*/

class PositionBarComponent extends Component {

  constructor(props) {
    super(props);
    this.canvas = createRef();
    this.draw = throttle(this.draw, this.props.viewpoint.msecsPerFps);
  }

  componentDidMount() {
    this.ctx = new Canvas(this.canvas.current);
    this.draw();
  }

  componentDidUpdate() {
    this.draw();
  }

  draw() {
    this.ctx.startDrawingFrame();
    this.ctx.font(this.props.font);

    const [tileWidth, tileHeight] = this.props.viewpoint.tileSizes;
    const yPos = 0;
    const startTile = Math.floor(this.props.position.xPos / tileWidth) - 1 + this.props.startIndex;
    const tiles = Math.ceil(this.props.viewpoint.width / tileWidth) + 1;
    let xPos = -this.props.position.xPos % tileWidth;
    for (let i = startTile; i < (startTile + tiles); i++) {
      if (i % this.props.markerSteps === 0) {
        this.ctx.fillText(i, xPos, yPos, tileWidth, tileHeight);
      } else {
        this.ctx.fillText(".", xPos, yPos, tileWidth, tileHeight);
      }
      xPos += this.props.viewpoint.tileSizes[0];
    }
    this.ctx.endDrawingFrame();
  }

  render() {
    const style = {
      display: "block",
    };
		const height = this.props.height * 1 + 2;
    return (
      <canvas
        ref={this.canvas}
        width={this.props.viewpoint.width}
        height={height}
				style={style}
      />
    );
  }
}

PositionBarComponent.defaultProps = {
  font: "12px Arial",
  height: 15,
  markerSteps: 2,
  startIndex: 1,
};

PositionBarComponent.PropTypes = {
  /**
   * Font of the sequence labels, e.g. `20px Arial`
   */
  font: PropTypes.string,

  /**
   * Height of the PositionBar (in pixels), e.g. `100`
   */
  height: PropTypes.number,

  /**
   * At which steps the position labels should appear, e.g. `2` for (1, 3, 5)
   */
  markerSteps: PropTypes.number,

  /**
   * At which number the PositionBar marker should start counting.
   * Typical values are: `1` (1-based indexing) and `0` (0-based indexing).
   */
  startIndex: PropTypes.number,
};

const mapStateToProps = state => {
  return {
    position: state.position,
    viewpoint: state.viewpoint,
    maxLength: state.sequences.maxLength,
  }
};

export default msaConnect(
  mapStateToProps,
)(PositionBarComponent);
