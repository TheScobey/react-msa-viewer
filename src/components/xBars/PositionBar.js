/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  pick
} from 'lodash-es';

import msaConnect from '../../store/connect'

import XBar from './xBar';

function createMarker({markerSteps, startIndex, tileWidth, font, markerComponent}) {
  /**
   * Displays an individual sequence name.
   */
  class Marker extends PureComponent {
    render() {
      const {index, ...otherProps} = this.props;
      if (markerComponent) {
        const MarkerComponent = markerComponent;
        return <MarkerComponent index={index} />
      } else {
        otherProps.style = {
          width: tileWidth,
          display: "inline-block",
          textAlign: "center",
        }
        let name;
        if (index % markerSteps === 0) {
          name = index+ 0 + startIndex;
        } else {
          name = '.';
        }
        return (
          <div {...otherProps}>
            {name}
          </div>
        );
      }
    }
  }
  return Marker;
}

/**
* Displays the sequence names with an arbitrary Marker component
*/
class HTMLPositionBarComponent extends PureComponent {

  componentWillMount() {
    this.updateMarker();
  }

  componentWillUpdate() {
    this.updateMarker();
  }

  updateMarker() {
    this.marker = createMarker(pick(this.props, [
      "markerSteps", "startIndex", "tileWidth", "markerComponent"
    ]));
  }

  render() {
    const {cacheElements,
      markerSteps,
      startIndex,
      dispatch,
      markerComponent,
      ...otherProps} = this.props;
    return (
      <XBar
        tileComponent={this.marker}
        cacheElements={cacheElements}
        {...otherProps}
      />
    );
  }
}

HTMLPositionBarComponent.defaultProps = {
  style: {
    font: "12px Arial",
  },
  height: 15,
  markerSteps: 2,
  startIndex: 1,
  cacheElements: 10,
};

HTMLPositionBarComponent.propTypes = {
  /**
   * Font of the sequence labels, e.g. `20px Arial`
   */
  font: PropTypes.string,

  /**
   * Height of the PositionBar (in pixels), e.g. `100`
   */
  height: PropTypes.number,

  /**
   * At which steps the position labels should appear, e.g. `2` for (1, 3, 5)
   */
  markerSteps: PropTypes.number,

  /**
   * At which number the PositionBar marker should start counting.
   * Typical values are: `1` (1-based indexing) and `0` (0-based indexing).
   */
  startIndex: PropTypes.number,

  /**
   * Component to create markers from.
   */
  markerComponent: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
}

const mapStateToProps = state => {
  return {
    sequences: state.sequences.raw,
    maxLength: state.sequences.maxLength,
    width: state.props.width,
    tileWidth: state.props.tileWidth,
    nrXTiles: state.sequenceStats.nrXTiles,
  }
}

export default msaConnect(
  mapStateToProps,
)(HTMLPositionBarComponent);
