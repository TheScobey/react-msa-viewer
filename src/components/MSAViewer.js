/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import MSAProvider from '../store/provider';

import PositionBar from './PositionBar';
import SequenceViewer from './SequenceViewer';
import SequenceOverview from './SequenceOverview';
import OverviewBar from './OverviewBar';
import Labels from './Labels';

import createMSAStore from '../store/createMSAStore';
import propsToRedux from '../store/propsToRedux';

const labelsAndSequenceDiv = {
  display: "flex",
};

// TODO: support changing the store dynamically
// TODO: when props of children update -> update store
// TODO: support using the child components in stand-alone mode
class MSAViewerComponent extends Component {

  componentWillMount() {
    //this.store = this.props.store || createMSAStore(
      //pick(this.props, MSAViewerProps)
    //);
  }
  render() {
    const {children, msaStore, ...otherProps} = this.props;
    if (children) {
      console.log("children: ", children);
      return (
        <MSAProvider store={msaStore}>
          <div {...otherProps}>
            {children}
          </div>
        </MSAProvider>
      );
    } else {
      // TODO: add more advanced layouts
      const currentState = msaStore.getState();
      const labelsPadding = currentState.viewpoint.tileSizes[1];
      const overviewBarHeight = currentState.viewpoint.overviewBar.height;
      const labelsStyle = {
        paddingTop: labelsPadding + overviewBarHeight,
      }
      const separatorPadding = {
        height: 10,
      };
      return (
        <MSAProvider store={msaStore}>
          <div style={labelsAndSequenceDiv}>
            <Labels
              style={labelsStyle}
            />
            <div>
              <OverviewBar />
              <PositionBar />
              <SequenceViewer />
              <div style={separatorPadding} />
              <SequenceOverview />
            </div>
          </div>
        </MSAProvider>
      );
    }
  }
}

const MSAViewer = propsToRedux(MSAViewerComponent);

MSAViewer.PropTypes = {
  // store
  // sequences, required
};

// TODO: re-include propsToRedux here?
export default MSAViewer;
export {
  createMSAStore,
  Labels,
  MSAViewer,
  OverviewBar,
  PositionBar,
  SequenceOverview,
  SequenceViewer,
};
