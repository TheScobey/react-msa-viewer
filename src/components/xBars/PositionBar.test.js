/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React from 'react';
import renderer from 'react-test-renderer';

import dummySequences from '../../test/dummySequences';
import { PositionBar } from './PositionBar';
import MSAViewer from '../MSAViewer';

it('renders properly', () => {
  const component = renderer.create(
    <MSAViewer sequences={[...dummySequences]}>
      <PositionBar nrXTiles={5} tileWidth={20} maxLength={20} />
    </MSAViewer>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
