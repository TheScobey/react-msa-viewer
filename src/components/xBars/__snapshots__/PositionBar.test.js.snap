// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div>
  <div
    height={15}
    style={
      Object {
        "font": "12px Arial",
      }
    }
  >
    <div
      style={
        Object {
          "overflow": "hidden",
          "position": "relative",
          "whiteSpace": "nowrap",
          "width": undefined,
        }
      }
    >
      <div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          1
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          3
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          5
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          7
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          9
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          11
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          13
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          15
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          17
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          19
        </div>
        <div
          style={
            Object {
              "display": "inline-block",
              "textAlign": "center",
              "width": 20,
            }
          }
        >
          .
        </div>
      </div>
    </div>
  </div>
</div>
`;
